dataset/groupby.cpp

```diff
@@ -369,10 +390,18 @@ GroupBy<Dataset> groupby(const DatasetConstView &dataset, const Dim dim) {
 GroupBy<Dataset> groupby(const DatasetConstView &dataset, const Dim dim,
                          const VariableConstView &bins) {
   const auto &key = dataset.coords()[dim];
-  return {
-      dataset,
-      core::CallDType<double, float, int64_t, int32_t>::apply<MakeBinGroups>(
-          key.dtype(), key, bins)};
+  return callDType<Dataset, DatasetConstView>(dataset, key, bins);
+}
+
+/// Create GroupBy<Dataset> object as part of "split-apply-combine" mechanism.
+///
+/// Groups the slices of `dataset` according to values in given by a coord.
+/// Grouping of a coord is according to given `bins`, which will be added as a
+/// new coordinate to the output in a later apply/combine step.
+GroupBy<Dataset> groupby(const DatasetConstView &dataset,
+                         const VariableConstView &variable,
+                         const VariableConstView &bins) {
```

Comment:
Should there be a check here that `variable` has the correct dimensions?

Reply:
But we don't check `Dim dim` dimensions for similar calls, do we? Should we?

Reply:
No, because in the old overload the key is a coord of the data array or dataset. This ensures that it is "valid". If we now support a key that is *not* a coord, we wil first need to check if it would be valid to add it as a coord.

Reply:
I suppose adding a unit test with a bad variable as key will show the issue: https://github.com/scipp/scipp/blob/c97e165cf9cbe96a7c7a983def350148cb45a5de/dataset/groupby.cpp#L294 just assumes that the key has the correct dim and length and setups `groups`, which is later used to slice the data array or dataset.